src/worker.rs

```diff
@@ -60,74 +60,73 @@ impl<'a> SidekiqWorker<'a> {
-        let clock = tick(Duration::from_secs(1));
+        let mut last_sync = Instant::now();
         loop {
-            select! {
-                default => {
-                    debug!("{} run queue once", self.id);
-                    match self.run_queue_once() {
-                        Ok(true) => self.processed += 1,
-                        Ok(false) => {}
-                        Err(e) => {
-                            self.failed += 1;
-                            warn!("uncaught error '{}'", e);
-                        }
-                    };
-                },
-                recv(clock) -> _ => {
-                    // synchronize state
-                    debug!("{} syncing state", self.id);
-                    self.sync_state();
-                    debug!("{} syncing state done", self.id);
-                },
-                recv(rx) -> op => {
-                    if let Ok(Operation::Terminate) = op {
-                        info!("{} Terminate signal received, exiting...", self.id);
-                        self.tx.send(Signal::Terminated(self.id.clone())).ok();
-                        debug!("{} Terminate signal sent", self.id);
-                        return;
-                    } else {
-                        unimplemented!()
-                    }
-                },
+            match self.rx.try_recv() {
+                Ok(Operation::Terminate) => {
+                    info!("{} Terminate signal received, exiting...", self.id);
+                    self.tx.send(Signal::Terminated(self.id.clone())).await.ok();
+                    debug!("{} Terminate signal sent", self.id);
+                    return;
+                }
+                _ => (),
+            }
+
+            // Note: this blocks for 1 second if no job is available in Redis
+            match self.run_queue_once().await {
+                Ok(true) => self.processed += 1,
+                Ok(false) => (),
+                Err(e) => {
+                    self.failed += 1;
+                    warn!("uncaught error '{}'", e);
+                }
+            }
+
+            if last_sync.elapsed().as_secs() >= 1 {
+                last_sync = Instant::now();
+                debug!("{} syncing state", self.id);
+                self.sync_state().await;
+                debug!("{} syncing state done", self.id);
             }
         }
     }
 
-
-    fn run_queue_once(&mut self) -> Result<bool> {
+    async fn run_queue_once(&mut self) -> Result<bool> {
         let queues: Vec<String> = self.queues.iter().map(|q| self.queue_name(q)).collect();
 
-        let result: Option<(String, String)> = self.pool.get()?.brpop(queues, 10)?;
+        let mut result: Result<Option<(String, String)>> = Ok(None);
+        task::block_in_place(|| {
+            Handle::current().block_on(async {
+                result = self.redis.brpop(queues, 1).await.map_err(From::from);
+            });
+        });
+        let result = result?;
```

Comment:
Internally, [`brpop` uses `query_async`](https://docs.rs/redis/latest/src/redis/commands/macros.rs.html#163-170). What's the purpose of doing this, given that the call wouldn't block the event queue anyways?

Reply:
This is all black magic to me so I wouldn't know. `brpop` is a blocking call even if the function is async. The future here doesn't return until a result is found in Redis or the timeout is reached.

The [block_in_place docs](https://docs.rs/tokio/latest/tokio/task/fn.block_in_place.html) suggest this pattern for async code that is blocking 🤷‍♀️

Reply:
> `brpop` is a blocking call even if the function is async. The future here doesn't return until a result is found in Redis or the timeout is reached.

What do you mean by this? An `async` function being `await`ed will not return until the future completes, regardless of whether it's wrapped in `block_in_place`. "Blocking" in the doc link refers to synchronous blocking code like the synchronous (non-`async`) version of [`br_pop`](https://docs.rs/redis/latest/src/redis/commands/mod.rs.html#66-18).

The current code here will not reach line 116 until the future on line 113 resolves. Did you want this to run in the background instead, and have the code continue with lines 116 and onwards in the meantime?

Reply:
Like I said, this is black magic to me. I can't find any documentation that explains exactly what happens for an async command that blocks on IO. The documentation covers non-async blocking commands.

Reply:
>The current code here will not reach line 116 until the future on line 113 resolves. Did you want this to run in the background instead, and have the code continue with lines 116 and onwards in the meantime?

No, the blocking behavior here is desired. I'm just trying to prevent starving the tokio runtime.

Reply:
Ah, so to be clear, `br_pop` isn't an async function that blocks synchronously on I/O. It's an asynchronous function that we can freely `await` on. When it comes to I/O in async code, it's internally done asynchronously, yielding to the event loop as necessary. This means there's no such thing really as async functions that block on I/O.

An `async` function that performs work by synchronously blocking internally is almost definitely a bug 😄

Reply:
Okay that's what I would assume, but the documentation isn't clear in this case so I figured I'd go ahead and do what the documentation suggests for some kinds of "blocking" operations.

I'll remove the `block_in_place` call.